### ISSUE TITLE
Failed to embed BZip2 from DotNetZip as Compressor

### ISSUE THREAD

**Issue by dmitrii-lobanov-1982** [benfoster/Fabrik.Common#14]
I've tried to use BZip2OutputStream from Ionic.BZip2 (DotNetZip) to implement my own compressor and I failed. I've ended up with output stream contains only 4 bytes and then the error occurs on client when receiving the answer.

I've implemented compressor as direct implementation of ICompressor interface, here's the method for compression:

```
public async Task Compress(Stream source, Stream compressed)
{
    using (BZip2OutputStream bzip = new BZip2OutputStream(compressed, true))
    {
        await source.CopyToAsync(bzip);
    }
}
```

BTW Ionic.Zlib.GZipStream works well instead of .NET GZipStream.

**Comment by dmitrii-lobanov-1982** [comment]
Looks like there're some issues with gzip2 algorithm itself and how it works with underlying stream. because the following code works fine:

```
public async Task Compress(Stream source, Stream destination)
{
    using (MemoryStream mem = new MemoryStream())
    {
        using (BZip2OutputStream bzip = new BZip2OutputStream(mem, true))
        {
            await source.CopyToAsync(bzip);
        }

        mem.Seek(0, SeekOrigin.Begin);
        await mem.CopyToAsync(destination);
    }
}
```

**Comment by benfoster** [comment]
Yes you will need to reset the stream position after reading it. This isn't
necessary with the built in GZipStream due to how it is being initialized
(I think - it's been a while since I wrote this code but I do remember
running into this issue before).

On 20 February 2013 12:08, Dmitry Lobanov notifications@github.com wrote:

> Looks like there're some issues with gzip2 algorithm itself and how it
> works with underlying stream. because the following code works fine:
> 
> public async Task Compress(Stream source, Stream destination)
> {
>     using (MemoryStream mem = new MemoryStream())
>     {
>         using (BZip2OutputStream bzip = new BZip2OutputStream(mem, true))
>         {
>             await source.CopyToAsync(bzip);
>         }
> 
> ```
>     mem.Seek(0, SeekOrigin.Begin);
>     await mem.CopyToAsync(destination);
> }
> ```
> 
> }
> 
>  —
> Reply to this email directly or view it on GitHubhttps://github.com/benfoster/Fabrik.Common/issues/14#issuecomment-13828617.

## 

Ben Foster

planetcloud
Rose Cottage, The Lane
Cotham
Nottinghamshire
NG23 5JT

www.planetcloud.co.uk

**
_fabrik._
Our portfolio and blogging platform for creatives
info@getfabrik.com
www.getfabrik.com
